### ISSUE TITLE
How to package wisp as a fat jar?

### ISSUE THREAD

**Issue by cacti77** [quantifind/wisp#37]
I want to use wisp in Apache Zeppelin behind a corporate firewall but am having terrible trouble trying to fetch wisp's dependencies through the corporate proxy:(. So as a work-around I would like to package wisp and all its dependencies into a single fat/uber jar.

I've tried reading through the [notes on sbt assembly](https://github.com/sbt/sbt-assembly), hacking with wisp/project/Build.scala and adding `addSbtPlugin("com.eed3si9n" % "sbt-assembly" % "0.11.2")` to project/plugins.sbt but I can't even get Build.scala to compile with the changes I've made. Unfortunately I'm new to Scala and struggling with this. It would be great if you could please add the ability to package wisp as a fat jar, or explain here how I could do it. Thanks very much!

**Comment by cacti77** [comment]
Ok, I might have answered my own question here...

In addition to the change above made in project/plugins.sbt I made these changes to project/Build.scala (changes highlighted with XXX):

```
import annotation.tailrec

// XXX
import sbtassembly.Plugin._
import AssemblyKeys._


object WispBuild extends Build {

	lazy val wisp = Project("wisp", file("core"), settings = wispSettings)

	// XXX
	val meta = """META.INF(.)*""".r
	mergeStrategy in assembly := {
		case PathList("javax", "servlet", xs @ _*) => MergeStrategy.first
		case PathList(ps @ _*) if ps.last endsWith ".html" => MergeStrategy.first
		case n if n.startsWith("reference.conf") => MergeStrategy.concat
		case n if n.endsWith(".conf") => MergeStrategy.concat
		case meta(_) => MergeStrategy.discard
		case x => MergeStrategy.first
	}

	def sharedSettings = Defaults.defaultSettings ++
		ReleasePlugin.releaseSettings ++
		BranchRelease.branchSettings ++
		xerial.sbt.Sonatype.sonatypeSettings ++
		assemblySettings ++ // XXX
		Seq(
			scalaVersion := "2.11.6",
			crossScalaVersions := Seq("2.10.5", "2.11.6"),
			organization := "com.quantifind",
			scalacOptions := Seq("-deprecation", "-unchecked", "-optimize"),
			retrieveManaged := true,
			transitiveClassifiers in Scope.GlobalScope := Seq("sources"),
			resolvers ++= Seq(
				"sonatype-snapshots" at "http://oss.sonatype.org/content/repositories/snapshots",
				"sonatype-releases" at "http://oss.sonatype.org/content/repositories/releases"
			),
```
Then I ran `sbt 'project wisp' assembly` in the `wisp` parent folder and wisp/core/target/scala-2.11/Wisp-assembly-0.0.4.jar was eventually built.  For the record, these were the jars that were included during the assembly process:

```
[info] Including: unfiltered-util_2.11-0.8.3.jar
[info] Including: commons-codec-1.4.jar
[info] Including: unfiltered-jetty_2.11-0.8.3.jar
[info] Including: jetty-webapp-8.1.13.v20130916.jar
[info] Including: jetty-xml-8.1.13.v20130916.jar
[info] Including: jetty-util-8.1.13.v20130916.jar
[info] Including: scala-library-2.11.6.jar
[info] Including: jetty-servlet-8.1.13.v20130916.jar
[info] Including: jetty-security-8.1.13.v20130916.jar
[info] Including: jetty-server-8.1.13.v20130916.jar
[info] Including: javax.servlet-3.0.0.v201112011016.jar
[info] Including: jetty-continuation-8.1.13.v20130916.jar
[info] Including: jetty-http-8.1.13.v20130916.jar
[info] Including: jetty-io-8.1.13.v20130916.jar
[info] Including: sumac_2.11-0.3.0.jar
[info] Including: commons-math3-3.4.1.jar
[info] Including: commons-io-2.4.jar
[info] Including: json4s-jackson_2.11-3.2.10.jar
[info] Including: json4s-core_2.11-3.2.10.jar
[info] Including: json4s-ast_2.11-3.2.10.jar
[info] Including: scala-reflect-2.11.0.jar
[info] Including: paranamer-2.6.jar
[info] Including: scalap-2.11.0.jar
[info] Including: scala-xml_2.11-1.0.1.jar
[info] Including: scala-parser-combinators_2.11-1.0.1.jar
[info] Including: jackson-databind-2.3.1.jar
[info] Including: jackson-annotations-2.3.0.jar
[info] Including: jackson-core-2.3.1.jar
[info] Including: unfiltered-filter_2.11-0.8.3.jar
[info] Including: unfiltered_2.11-0.8.3.jar
[info] Including: scala-compiler-2.11.0.jar
```
Unfortunately I still can't get the fat jar to work in Zeppelin so, unless I packaged it incorrectly, there may be another problem to solve...

**Comment by AustinBGibbons** [comment]
Hey cacti77, I have never played with Apache Zeppelin myself, but from the look of it, it is a standalone charting tool - can you describe why you want to use wisp within zeppelin? On the surface of things, it does not look like they should be combined.

Additionally Wisp starts up a small webserver and serves images through highcharts. It hyperlinks to the highcharts code - it does not include a copy of the highcharts source, so that could be another problem for your firewall.

Best!
Austin

**Comment by cacti77** [comment]
Hi @AustinBGibbons; thanks for replying.  Zeppelin is good for interfacing with Spark but it has very limited plotting capabilities. That's a gap which wisp fills very nicely since it bridges from Scala to Highcharts!

Starting up a small webserver shouldn't be a problem. I can probably just change the wisp code to capture the HTML string that would otherwise be sent to the webserver and hopefully get Zeppelin to plot that instead, preferably in-line in a notebook 'paragraph'.

Unfortunately Zeppelin can't punch through a proxy right now - see https://issues.apache.org/jira/browse/ZEPPELIN-169 . So you're probably right about Highcharts. So my current plan is to build a fat jar of wisp and then `z.load()` it from the local file-system using Zeppelin's [dependency loader](https://zeppelin.incubator.apache.org/docs/interpreter/spark.html).  I guess I'll have to include Highcharts in the fat jar too if it's to work behind the proxy...

**Comment by cacti77** [comment]
Correcting myself, I shouldn't need to include Highcharts in the jar!! I see it's a Javascript library that you're loading from a script element in the web page you construct in HighchartsImpl.scala. Since I can execute `println(scala.io.Source.fromURL("http://code.highcharts.com/4.0.4/highcharts.js").mkString)` successfully in Zeppelin behind my proxy I don't think fetching Highcharts should be a problem.

**Comment by cacti77** [comment]
However, in Zeppelin, even though I can `z.load()` my fat jar for wisp and run `import com.quantifind.charts.Highcharts._` I get this error when I try running `line((1 to 10), (1 to 10))`:
```
java.lang.NoSuchMethodError: scala.Predef$.ArrowAssoc(Ljava/lang/Object;)Ljava/lang/Object;
	at com.quantifind.sumac.FieldArgs$class.registerAnnotationValidationUpdate(FieldArgs.scala:94)
	at com.quantifind.charts.repl.PlotServer$$anon$1.registerAnnotationValidationUpdate(PlotServer.scala:39)
	at com.quantifind.sumac.FieldArgs$class.registerAnnotationValidation(FieldArgs.scala:98)
	at com.quantifind.charts.repl.PlotServer$$anon$1.registerAnnotationValidation(PlotServer.scala:39)
	at com.quantifind.sumac.FieldArgs$class.$init$(FieldArgs.scala:106)
	at com.quantifind.charts.repl.PlotServer$$anon$1.<init>(PlotServer.scala:39)
	at com.quantifind.charts.repl.PlotServer.parseArgs(PlotServer.scala:39)
	at com.quantifind.charts.repl.WebPlot$class.startWispServer(PlotLike.scala:148)
	at com.quantifind.charts.Highcharts$.startWispServer(Highcharts.scala:18)
	at com.quantifind.charts.repl.WebPlot$class.$init$(PlotLike.scala:81)
	at com.quantifind.charts.Highcharts$.<init>(Highcharts.scala:18)
	at com.quantifind.charts.Highcharts$.<clinit>(Highcharts.scala)
	at $iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC.<init>(<console>:25)
	at $iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC.<init>(<console>:30)
	at $iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC.<init>(<console>:32)
	at $iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC.<init>(<console>:34)
	at $iwC$$iwC$$iwC$$iwC$$iwC$$iwC.<init>(<console>:36)
	at $iwC$$iwC$$iwC$$iwC$$iwC.<init>(<console>:38)
	at $iwC$$iwC$$iwC$$iwC.<init>(<console>:40)
	at $iwC$$iwC$$iwC.<init>(<console>:42)
	at $iwC$$iwC.<init>(<console>:44)
	at $iwC.<init>(<console>:46)
	at <init>(<console>:48)
	at .<init>(<console>:52)
	at .<clinit>(<console>)
	at .<init>(<console>:7)
	at .<clinit>(<console>)
	at $print(<console>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.spark.repl.SparkIMain$ReadEvalPrint.call(SparkIMain.scala:1065)
	at org.apache.spark.repl.SparkIMain$Request.loadAndRun(SparkIMain.scala:1338)
	at org.apache.spark.repl.SparkIMain.loadAndRunReq$1(SparkIMain.scala:840)
	at org.apache.spark.repl.SparkIMain.interpret(SparkIMain.scala:871)
	at org.apache.spark.repl.SparkIMain.interpret(SparkIMain.scala:819)
	at org.apache.zeppelin.spark.SparkInterpreter.interpretInput(SparkInterpreter.java:600)
	at org.apache.zeppelin.spark.SparkInterpreter.interpret(SparkInterpreter.java:576)
	at org.apache.zeppelin.spark.SparkInterpreter.interpret(SparkInterpreter.java:569)
	at org.apache.zeppelin.interpreter.ClassloaderInterpreter.interpret(ClassloaderInterpreter.java:57)
	at org.apache.zeppelin.interpreter.LazyOpenInterpreter.interpret(LazyOpenInterpreter.java:93)
	at org.apache.zeppelin.interpreter.remote.RemoteInterpreterServer$InterpretJob.jobRun(RemoteInterpreterServer.java:277)
	at org.apache.zeppelin.scheduler.Job.run(Job.java:170)
	at org.apache.zeppelin.scheduler.FIFOScheduler$1.run(FIFOScheduler.java:118)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
```
Have you seen this before, Austin? Do you think something is still missing from my fat jar??

**Comment by AustinBGibbons** [comment]
Hey I am out of town and won't be able to try for myself until Monday, but it looks like you are running some code from one of the dependencies which is great! But it's missing a scala predef function?




Just to check, are you using the same version of scala in Zeppelin as the one you used when assembling the wisp jar?



—

**Comment by cacti77** [comment]
Thank you; you're absolutely right! We discovered this independently yesterday too. When I ran `util.Properties.versionString` in Zeppelin 0.5.0 it reported version 2.10.4 of Scala. By default `sbt 'project wisp' assembly` is compiling using Scala 2.11.6, as defined in your Build.scala file. But, luckily for me, it also cross-compiles against 2.10.5: `crossScalaVersions := Seq("2.10.5", "2.11.6")`.

So, after looking [here](http://www.scala-sbt.org/0.13/docs/Cross-Build.html#Cross-Building+a+Project), I just did `sbt 'project wisp' +assembly` to build the fat jars for both Scala versions.  Then I imported wisp/core/target/__scala-2.10__/Wisp-assembly-0.0.4.jar into Zeppelin using `z.load()` and then I could successfully create Wisp plots from within Zeppelin - yippee!
